core/utils/ui_position.js

```diff
@@ -0,0 +1,195 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Utility functions for positioning UI elements.
+ * @author kozbial@google.com (Monica Kozbial)
+ */
+'use strict';
+
+goog.provide('Blockly.utils.uiPosition');
+
+goog.require('Blockly.Scrollbar');
+goog.require('Blockly.utils.Rect');
+goog.require('Blockly.utils.toolbox');
+
+goog.requireType('Blockly.MetricsManager');
+goog.requireType('Blockly.WorkspaceSvg');
+
+
+/**
+ * Enum for vertical positioning.
+ * @enum {number}
+ */
+Blockly.utils.uiPosition.verticalPositionType = {
+  TOP: 0,
+  CENTER: 1,
+  BOTTOM: 2
+};
+
+/**
+ * Enum for horizontal positioning.
+ * @enum {number}
+ */
+Blockly.utils.uiPosition.horizontalPositionType = {
+  LEFT: 0,
+  CENTER: 1,
+  RIGHT: 2
+};
+
+/**
+ * An object defining a horizontal and vertical positioning.
+ * @typedef {{
+ *            horizontal: !Blockly.utils.uiPosition.horizontalPositionType,
+ *            vertical: !Blockly.utils.uiPosition.verticalPositionType
+ *          }}
+ */
+Blockly.utils.uiPosition.PositionType;
+
+/**
+ * Enum for bump rules to use for dealing with collisions.
+ * @enum {number}
+ */
+Blockly.utils.uiPosition.bumpRule = {
+  BUMP_UP: 0,
+  BUMP_DOWN: 1,
+  BUMP_LEFT: 2,
+  BUMP_RIGHT: 3
```

Comment:
nit: since this is already a `bumpRule` enum you can probably change these to just be `UP`, `DOWN` etc.